OpenRA.Mods.Common/Traits/Render/Hovers.cs

```diff
@@ -13,41 +13,78 @@
 using System.Collections.Generic;
 using System.Drawing;
 using System.Linq;
+using OpenRA;
 using OpenRA.Graphics;
 using OpenRA.Traits;
 
 namespace OpenRA.Mods.Common.Traits.Render
 {
 	[Desc("Changes the visual Z position periodically.")]
-	public class HoversInfo : ConditionalTraitInfo, Requires<IMoveInfo>
+	public class HoversInfo : PausableConditionalTraitInfo, Requires<IMoveInfo>
 	{
 		[Desc("Amount of Z axis changes in world units.")]
 		public readonly int OffsetModifier = -43;
 
 		public readonly int MinHoveringAltitude = 0;
 
+		[Desc("Amount of ticks it takes to reach OffsetModifier.")]
+		public readonly int Ticks = 25;
+
+		[Desc("Amount of ticks it takes to fall to the ground from the highest point when disabled.")]
+		public readonly int FallTicks = 12;
+
+		[Desc("Amount of ticks it takes to rise from the ground to InitialHeight.")]
+		public readonly int RiseTicks = 17;
+
+		public readonly int InitialHeight = 384;
+
 		public override object Create(ActorInitializer init) { return new Hovers(this, init.Self); }
 	}
 
-	public class Hovers : ConditionalTrait<HoversInfo>, IRenderModifier
+	public class Hovers : PausableConditionalTrait<HoversInfo>, IRenderModifier, ITick
 	{
 		readonly HoversInfo info;
+		readonly int stepPercentage;
+		readonly int fallTickHeight;
+
+		int ticks = 0;
+		WVec worldVisualOffset = WVec.Zero;
 
 		public Hovers(HoversInfo info, Actor self)
 			: base(info)
 		{
 			this.info = info;
+			this.stepPercentage = 256 / info.Ticks;
+			this.fallTickHeight = (info.InitialHeight + info.OffsetModifier) / info.FallTicks;
+		}
+
+		void ITick.Tick(Actor self)
+		{
+			if (IsTraitDisabled || IsTraitPaused)
+			{
+				if (worldVisualOffset.Z < 0)
+					return;
+
+				var fallTicks = worldVisualOffset.Z / fallTickHeight - 1;
+				worldVisualOffset = new WVec(0, 0, fallTickHeight * fallTicks);
+			}
+			else
+				ticks++;
 		}
 
 		IEnumerable<IRenderable> IRenderModifier.ModifyRender(Actor self, WorldRenderer wr, IEnumerable<IRenderable> r)
 		{
-			if (self.World.Paused || IsTraitDisabled)
-				return r;
+			if (!IsTraitDisabled && !IsTraitPaused)
+			{
+				var visualOffset = self.World.Map.DistanceAboveTerrain(self.CenterPosition).Length >= info.MinHoveringAltitude
+					? new WAngle(ticks % (info.Ticks * 4) * stepPercentage).Sin() : 0;
+				var currentHeight = info.OffsetModifier * visualOffset / 1024 + info.InitialHeight;
+
+				if (worldVisualOffset.Z < currentHeight)
```

Comment:
Can you please add a comment just above this explaining that this allows the actor to rise up to normal level over a few ticks after being disabled (I think?).